todo_app/app.py

```diff
@@ -3,29 +3,108 @@
+    CLIENT_ID = os.environ.get("GITHUB_CLIENT_ID", None)
+    CLIENT_SECRET = os.environ.get("GITHUB_CLIENT_SECRET", None)
+    SECRET_KEY = os.environ.get("SECRET_KEY", None)
+    WRITER_ROLE = os.environ.get("ROLEWRITER_USER", None)  
+    redirect_uri=os.environ.get("GITHUB_REDIRECT_URI", None)
+    app.secret_key = SECRET_KEY   
+    client = WebApplicationClient(CLIENT_ID)
+    login_manager = LoginManager()
+    login_manager.init_app(app)
+          
+    @login_manager.unauthorized_handler
+    def unauthenticated():        
+        identity_url=client.prepare_request_uri('https://github.com/login/oauth/authorize',redirect_uri)      
+        return redirect(identity_url)
+
+    @app.route("/login/callback")
+    def callback():    
+        code = request.args.get("code")           
+        token_url, headers, body = client.prepare_token_request(
+            "https://github.com/login/oauth/access_token",
+            authorization_response=request.url,
+            redirect_url=request.base_url,
+            code=code,
+        )
+        token_response = requests.post(
+            token_url,
+            headers=headers,
+            data=body,
+            auth=(CLIENT_ID, CLIENT_SECRET),
+        )
+
+        # Parse the tokens!
+        params=client.parse_request_body_response(token_response.text)
+        uri, headers, body = client.add_token("https://api.github.com/user")
+        
+        userinfo_response = requests.get(uri, headers=headers, data=body)
+        
+        users_name = userinfo_response.json()["login"]
+        users_id = userinfo_response.json()["id"]        
+        email = userinfo_response.json()["email"]             
+        user = User(users_id,users_name,email,ROLES['reader'])
+        appsession['users_id'] = users_id
+        appsession['users_name'] = users_name   
+        appsession['email'] = email
+        if users_name == WRITER_ROLE:
+            user.access=ROLES['writer']
+        else:
+            user.access=ROLES['reader']
+
+        appsession['roles'] = user.access    
+        login_user(user)
+        return redirect(url_for('index'))
+
+     
+
+    @login_manager.user_loader
+    def load_user(user_id):         
+        m_user_id = appsession.get('users_id')
+        m_users_name = appsession.get('users_name') 
+        m_users_roles = appsession.get('roles')  
+        m_email = appsession.get('email')      
+        user = User(m_user_id,m_users_name,m_email,m_users_roles)         
+        return user
+
+   
     
     @app.route('/')
+    @login_required
+    @requires_roles('reader','writer')
     def index():             
         items = session.Boards().get_items()
-        item_view_model = session.ViewModel(items)
-        return render_template('index.html',view_model=item_view_model)
+        item_view_model = session.ViewModel(items)  
+        if current_user.is_anonymous:
+            mcurrent_user=''  
+            misWriter=True #for E2E testing
```

Comment:
This is fine, but if interested a nicer solution might be using [the anonymous user setting Flask Login provides](https://flask-login.readthedocs.io/en/latest/#anonymous-users)